test/Test_Grammar.re

```diff
@@ -18,7 +23,227 @@ let qcheck_map_annotation_test =
+        };
+
+        closure(closure_environment(Id.mk(), M.empty), empty_hole());
+      | Parens => parens(empty_hole())
+      | Cons => cons(empty_hole(), empty_hole())
+      | UnOp(op) => un_op(op, empty_hole())
+      | BinOp(op) => bin_op(op, empty_hole(), empty_hole())
+      | BuiltinFun => builtin_fun("string_compare")
+      | Match => match(empty_hole(), [])
+      | Cast => cast(empty_hole(), Typ.int(), Typ.string())
+      | ListConcat => list_concat(empty_hole(), empty_hole())
+      }
+    )
+  );
+};
+
+let sample_pattern = (cls_pat: Pat.cls): Grammar.UnitGrammar.pat => {
+  Grammar.UnitGrammar.(
+    Pat.(
+      switch (cls_pat) {
+      | Invalid => invalid("invalid")
+      | EmptyHole => empty_hole()
+      | MultiHole => multi_hole([Pat(empty_hole()), Pat(empty_hole())])
+      | Bool => bool(true)
+      | Int => int(1)
+      | Float => float(2.)
+      | String => string("hello")
+      | ListLit => list_lit([])
+      | Constructor => constructor("A", None)
+      | Var => var("x")
+      | Tuple => tuple([])
+      | Cons => cons(empty_hole(), empty_hole())
+      | Label => label("label")
+      | TupLabel => tup_label(label("label"), empty_hole())
+      | Parens => parens(empty_hole())
+      | Ap => ap(empty_hole(), empty_hole())
+      | Cast => cast(empty_hole(), Typ.int(), Typ.string())
+      | Wild => wild()
+      }
+    )
+  );
+};
+
+let sample_type = (cls_typ: Typ.cls): Grammar.UnitGrammar.typ => {
+  Grammar.UnitGrammar.(
+    Typ.(
+      switch (cls_typ) {
+      | Invalid => unknown(Hole(Invalid("invalid")))
+      | Int => int()
+      | Float => float()
+      | String => string()
+      | Bool => bool()
+      | List => list(unknown(Hole(EmptyHole)))
+      | Arrow => arrow(unknown(Hole(EmptyHole)), unknown(Hole(EmptyHole)))
+      | Var => var("x")
+      | Prod => prod([])
+      | TupLabel =>
+        tup_label(unknown(Hole(EmptyHole)), unknown(Hole(EmptyHole)))
+      | Parens => parens(unknown(Hole(EmptyHole)))
+      | Ap => ap(unknown(Hole(EmptyHole)), unknown(Hole(EmptyHole)))
+      | Rec => rec_(TPat.var("x"), unknown(Hole(EmptyHole)))
+      | Forall => forall(TPat.var("x"), unknown(Hole(EmptyHole)))
+      | EmptyHole => unknown(Hole(EmptyHole))
+      | SynSwitch => unknown(SynSwitch)
+      | Internal => unknown(Internal)
+      | Label => label("label")
+      | MultiHole => unknown(Hole(MultiHole([])))
+      | Sum => sum([])
+      | Constructor => assert(false) // Excluded because there is no Typ constructor
+      }
+    )
+  );
+};
+
+let sample_tpat = (cls_tpat: TPat.cls): Grammar.UnitGrammar.tpat => {
+  Grammar.UnitGrammar.(
+    TPat.(
+      switch (cls_tpat) {
+      | Invalid => invalid("invalid")
+      | EmptyHole => empty_hole()
+      | Var => var("x")
+      | MultiHole => multi_hole([TPat(empty_hole()), TPat(empty_hole())])
+      }
+    )
+  );
+};
+
 let tests = (
   "Grammar",
-  [QCheck_alcotest.to_alcotest(qcheck_map_annotation_test)],
+  [
+    test_case(
```

Comment:
I added these test cases to walk over all of the expressions, types, patterns, ...  to make sure that we can build one for each `cls` and the class is in 1-1 correspondence with the term types. This turns out not to be true for types for Constructor.


The tests help ensure cls are in order and that we have the factory functions. I think that's good enough for now.

Reply:
Maybe I'll find a way to derive both in the future.